### ISSUE TITLE
Warning inconsistencies between mac and ubuntu

### ISSUE THREAD

**Issue by LudvigOlsen** [lme4/lme4#540]
Hi,

When running a series of glmer models with different `control` argument settings, I noticed that I was getting a different number of warnings on mac and ubuntu, when using a formula, and specifying the control argument, such that warnings are expected. 

```
# Attach packages
library(cvms)
library(dplyr)
library(testthat)
library(lme4)
# also requires plyr to run

# A few helper functions first

# Finding current R version
check_R_version <- function(){
  major <- as.integer(R.Version()$major)
  minor <- as.numeric(strsplit(R.Version()$minor, ".", fixed = TRUE)[[1]][[1]])
  list("major"=major, "minor"=minor)
}

# Function for setting seed with 'Rounding' engine for backwards compatibility
set_seed_for_R_compatibility <- function(seed = 1) {
    version <- check_R_version()
    if ((version[["major"]] == 3 && version[["minor"]] >= 6) || version[["major"]] > 3) {
        args <- list(seed, sample.kind = "Rounding")
    } else {
        args <- list(seed)
    }
    suppressWarnings(do.call(set.seed, args))
}

# Get all lists in a list with a certain name
# Use: list_of_lists %c% 'list_name'
`%c%` <- function(x, n) lapply(x, `[[`, n)

# Now the example
 
# data from cvms
dat <- participant.scores %>% 
    dplyr::mutate(diagnosis = as.factor(diagnosis))

# Set the various variables 
formula <- "diagnosis ~ score + age + (1|session) + (1|age)"
family <- 'binomial'
control <- lme4::glmerControl(optimizer = "bobyqa")

# Fit 10 glmer models on random subsets of the data
# and collect the warnings and messages with evaluate_promise()

warnings_and_messages <- plyr::llply(1:10, function(s){
    process_ <- testthat::evaluate_promise({
        # set seed
        set_seed_for_R_compatibility(s)
        # fit the glmer model with a random subset of the data
        lme4::glmer(formula = formula, family = family, 
                             data = dplyr::sample_frac(dat, 0.95),
                             control = control)
        })
    list("warnings" = process_$warnings,
    "messages" = process_$messages)
})

# extract all the warnings and messages to each their variable
warns <- unlist(warnings_and_messages %c% "warnings")
mesgs <- unlist(warnings_and_messages %c% "messages")

# inspect the warnings
warns
```

Now this is where I get confused. On mac osx 10.12.6 I get **3** warnings:

```
[1] "Model failed to converge with max|grad| = 0.025616 (tol = 0.001, component 1)" 
[2] "Model failed to converge with max|grad| = 0.0161905 (tol = 0.001, component 1)"
[3] "Model failed to converge with max|grad| = 0.033809 (tol = 0.001, component 1)" 
```

and on ubuntu 18.04 I get the following **6** warnings:

```                                   
[1] "Model failed to converge with max|grad| = 0.0300621 (tol = 0.001, component 1)"                                      
[2] "Model failed to converge with max|grad| = 0.0231165 (tol = 0.001, component 1)"                                      
[3] "unable to evaluate scaled gradient"                                            
[4] "Model failed to converge: degenerate  Hessian with 1 negative eigenvalues"                                       
[5] "unable to evaluate scaled gradient"                                            
[6] "Model failed to converge: degenerate  Hessian with 1 negative eigenvalues"    
```

Now, I would never use these settings in the real world as the dataset is very small, but I'm wondering why there is a difference between the warnings thrown, given the same seeds?
Currently, I'm also seeing some differences with numeric results from the nnet package between the two platforms (edit: though only with a binomial model, not with the linear or multinom models). I have countless of glm(er) and lm(er) tests that are identical between the platforms though, so I'm wondering if it's just the (rather old) nnet package, that has that behaviour, or whether something is up with my mac.

Perhaps someone can try to replicate the difference in warnings?

## Edit 1
I had a friend run in on his mac, and also configured travis to test on mac as well. Both found only 3 warnings. So it seems that it is a general difference between mac and ubuntu. 
I forgot to mention that travis' ubuntu 16.04 agrees with my local ubuntu on the 6 warnings.

## Edit 2
I removed the REML part of the question, as @bbolker answered it, and it clutters the actual question about differences in warnings (REML is not an argument in glmer). 
The REML question that @bbolker answered below was:

> Another error I get when running `devtools::test()` and expecting those warnings is that the `‘REML’` messages are not recognized, because the `‘’` are instead `''`. So in the console (on both platforms btw.) REML is surrounded by `‘’`, and in test() it's `''`. Where this difference in the warning messages happen, I don't know.

Best,
Ludvig

**Comment by bbolker** [comment]
**quick**, incomplete answers:

- I'm not at all surprised to see slight differences in numerical outcomes, big enough to cause differences in warnings, from different platforms/compilers/etc.   I'll dig in a bit more when I get a chance.
- not sure why you're bothering to include `REML` with `glmer` (which doesn't take that argument)?  Clutters up the question a bit ...
- the differences in quotation marks across platforms are probably caused by differences in locale. On my Ubuntu 16.04 virtual machine:
```
> sQuote("a")
[1] "‘a’"
> Sys.getlocale()
[1] "LC_CTYPE=en_CA.UTF8;LC_NUMERIC=C;LC_TIME=en_CA.UTF8;LC_COLLATE=en_CA.UTF8;LC_MONETARY=en_CA.UTF8;LC_MESSAGES=en_CA.UTF8;LC_PAPER=en_CA.UTF8;LC_NAME=C;LC_ADDRESS=C;LC_TELEPHONE=C;LC_MEASUREMENT=en_CA.UTF8;LC_IDENTIFICATION=C"
> Sys.setlocale(locale="C")
[1] "LC_CTYPE=C;LC_NUMERIC=C;LC_TIME=C;LC_COLLATE=C;LC_MONETARY=C;LC_MESSAGES=en_CA.UTF8;LC_PAPER=en_CA.UTF8;LC_NAME=C;LC_ADDRESS=C;LC_TELEPHONE=C;LC_MEASUREMENT=en_CA.UTF8;LC_IDENTIFICATION=C"
> sQuote("a")
[1] "'a'"
```

**Comment by LudvigOlsen** [comment]
Hi Ben @bbolker!

Thanks for looking into it. :)

> not sure why you're bothering to include REML with glmer (which doesn't take that argument)? Clutters up the question a bit ...

Yeah, you're right. I think, I added it to check against an error that should always be there. And then I got the quotation error, and was curious about that as well as the difference in warnings.
I will remove it from the question to declutter it. 

> the differences in quotation marks across platforms are probably caused by differences in locale. On my Ubuntu 16.04 virtual machine

Perhaps it's best to manually set the locale when using testthat? That isn't a lme4 question, but if anyone has a point to make on that, it would be cool.
